### PR TITLE
add code line attribute for leading spaces #2011

### PR DESCRIPTION
**Pull request by bigld**
Use attribute abapgit-leadig-spaces to store leading spaces of coding lines at serialization. This is used in deserialization to restore formatting after XML normalization.
#2011

### CODE REVIEW

File: src/objects/zcl_abapgit_object_ssfo.clas.abap [hunk 1]
```diff
@@ -263,6 +97,120 @@ CLASS zcl_abapgit_object_ssfo IMPLEMENTATION.
...
+
+    ENDIF.
+
+    IF c_within_code_section = abap_true.
+      IF i_name = node_item.
+
+        TRY.
+            element ?= i_node.
+            leading_spaces = element->get_attribute_ns(
+                             EXPORTING name = zcl_abapgit_object_ssfo=>attrib_abapgit_leadig_spaces ).
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_ssfo.clas.abap]:
```suggestion
                              zcl_abapgit_object_ssfo=>attrib_abapgit_leadig_spaces ).
```

File: src/objects/zcl_abapgit_object_ssfo.clas.abap [hunk 2]
```diff
@@ -263,6 +97,120 @@ CLASS zcl_abapgit_object_ssfo IMPLEMENTATION.
...
+    IF c_within_code_section = abap_true.
+      IF i_name = node_item.
+
+        code_line = i_node->get_value( ).
+        "find 1st non space char
+        FIND FIRST OCCURRENCE OF REGEX '\S' IN code_line MATCH OFFSET offset.
+        IF sy-subrc = 0 AND offset > 0.
+          TRY.
+              element ?= i_node.
+              element->set_attribute( EXPORTING name  = attrib_abapgit_leadig_spaces
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_ssfo.clas.abap]:
```suggestion
              element->set_attribute( name  = attrib_abapgit_leadig_spaces
```

File: src/objects/zcl_abapgit_object_ssfo.clas.abap [hunk 3]
```diff
@@ -263,6 +97,120 @@ CLASS zcl_abapgit_object_ssfo IMPLEMENTATION.
...
+
+  METHOD handle_attrib_leading_spaces.
+
+    DATA element        TYPE REF TO if_ixml_element.
+    DATA leading_spaces TYPE string.
+    DATA coding_line    TYPE string.
+
+    IF i_name = zcl_abapgit_object_ssfo=>node_code.
+      c_within_code_section = abap_true.
+
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_ssfo.clas.abap]:
```suggestion
```

File: src/objects/zcl_abapgit_object_ssfo.clas.abap [hunk 4]
```diff
@@ -263,6 +97,120 @@ CLASS zcl_abapgit_object_ssfo IMPLEMENTATION.
...
+            element->set_value( coding_line ).
+
+          CATCH cx_sy_move_cast_error ##no_handler.
+        ENDTRY.
+
+      ELSEIF i_name <> node_code AND i_name <> node_text.
+        c_within_code_section = abap_false.
+
+      ENDIF.
+
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_ssfo.clas.abap]:
```suggestion
```

File: src/objects/zcl_abapgit_object_ssfo.clas.abap [hunk 5]
```diff
@@ -263,6 +97,120 @@ CLASS zcl_abapgit_object_ssfo IMPLEMENTATION.
...
+  METHOD set_attribute_leading_spaces.
+
+    DATA: element             TYPE REF TO if_ixml_element.
+    DATA: code_line           TYPE string.
+    DATA: offset              TYPE i.
+
+    IF i_name = zcl_abapgit_object_ssfo=>node_code.
+      c_within_code_section = abap_true.
+
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_ssfo.clas.abap]:
```suggestion
```

**Comment by christianguenter2** [comment]
Tested. LGTM.

**Comment by larshp** [comment]
cool, merging 👍

**Pull request merged**